sympy/core/compatibility.py

```diff
@@ -472,6 +472,13 @@ def default_sort_key(item, order=None):
     return (cls_index, 0, item.__class__.__name__
             ), args, S.One.sort_key(), S.One
 
+def _node_count(e):
+    if hasattr(e,'args'):
+        return 1 + sum(map(_node_count, e.args))
+    elif isinstance(e, int):
+        return 0
```

Comment:
We only call this `_node_count` function if `e` is an instance of `Basic`. That should mean that it is guaranteed to have `.args` that are also all `Basic`. The `hasattr` above shouldn't be necessary and it shouldn't be possible for `e` to be an `int`. I think these are both workarounds covering up bugs.

Can you show examples where the `hasattr` and the elif etc are needed?

Reply:
In the `stats` module there are many instances of `function` objects being in the `.args` list (Formerly these were counted as `0` in the preorder traversal of `e.count(Basic)` ) . An example is the core test of `stats\test\test_stochastic_process.py` 

```
sympy/stats/tests/test_stochastic_process.py:515: 
    assert P(X(t) < 1) == exp(-3*t)
sympy/stats/rv.py:809: in probability
    return Probability(condition, given_condition).doit(**kwargs)
sympy/stats/symbolic_probability.py:75: in doit
    return pspace(condition).probability(condition, given_condition,
sympy/stats/stochastic_process.py:53: in probability
    return self.process.probability(condition, given_condition, evaluate, **kwargs)
sympy/stats/stochastic_process_types.py:2104: in probability
    return _SubstituteRV._probability(condition, evaluate=evaluate, **kwargs)
sympy/stats/stochastic_process_types.py:1820: in _probability
    new_condition, new_givencondition = self._rvindexed_subs(condition, given_condition)
sympy/stats/stochastic_process_types.py:1755: in _rvindexed_subs
    expr = expr.subs(swapdict_expr)
sympy/core/basic.py:921: in subs
    k = list(ordered(sequence, default=False, keys=(
sympy/core/compatibility.py:589: in ordered
    d[f(a)].append(a)
sympy/core/basic.py:922: in <lambda>
    lambda x: -_nodes(x),
sympy/core/compatibility.py:492: in _nodes
    assert _node_count(e)==e.count(Basic)
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

e = <bound method PoissonProcess.distribution of PoissonProcess(X, 3)>

    def _node_count(e):
>       return 1 + sum(map(_node_count, e.args))
E       AttributeError: 'function' object has no attribute 'args'
```
Also there was an instance of an `int` object in the core test `unify\test\test_sympy.py` . This is a singular occurence and probably a bug in contrast to the `functions` module. 
```
    def test_FiniteSet_complex():
        from sympy import FiniteSet
        a, b, c, x, y, z = symbols('a,b,c,x,y,z')
>       expr = FiniteSet(Basic(1, x), y, Basic(x, z))

sympy/unify/tests/test_sympy.py:127: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/sets/sets.py:1778: in __new__
    for i in reversed(list(ordered(args))):
sympy/core/compatibility.py:608: in ordered
    yield from d[k]
sympy/core/compatibility.py:589: in ordered
    d[f(a)].append(a)
sympy/core/compatibility.py:492: in _nodes
    assert _node_count(e)==e.count(Basic)
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

e = 1

    def _node_count(e):
>       return 1 + sum(map(_node_count, e.args))
E       AttributeError: 'int' object has no attribute 'args'
```
Intrestingly I just found out returning `0` or `1` for either of the `function` or`int` objects didn't make any difference (all core tests passed in both cases). But the current `e.count(Basic)` returns them as `0` so probably that shoud be the way to go.

Reply:
There is also a few instances of `NoneType` being returned in 

```
>       assert P(Eq(Y[3], 2), Eq(Y[1], 1) & TransitionMatrixOf(Y, TO)).round(3) == Float(0.375, 3)

sympy/stats/tests/test_stochastic_process.py:90: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/core/decorators.py:266: in _func
    return func(self, other)
sympy/logic/boolalg.py:70: in __and__
    return And(self, other)
sympy/core/operations.py:489: in __new__
    _args = frozenset(cls._new_args_filter(args))
sympy/logic/boolalg.py:685: in _new_args_filter
    for x in ordered(args):
sympy/core/compatibility.py:607: in ordered
    yield from d[k]
sympy/core/compatibility.py:588: in ordered
    d[f(a)].append(a)
sympy/core/compatibility.py:491: in _nodes
    assert _node_count(e)==e.count(Basic)
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
sympy/core/compatibility.py:476: in _node_count
    return 1 + sum(map(_node_count, e.args))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

e = None

    def _node_count(e):
>       return 1 + sum(map(_node_count, e.args))
E       AttributeError: 'NoneType' object has no attribute 'args'
```
Again returning these cases as `1` or `0` still passed the test but `e.count(Basic)` returned them as `0`.

Reply:
Is that something like https://github.com/sympy/sympy/issues/20078#issuecomment-691712084

I think some part of this is being fixed in #20626

Reply:
This is why it's good that we don't cover over things like this so that these mistakes are picked up early.

Reply:
You're right, I'll remove the check.

#20626  this will fix the bugs in `stats` but the `int` instance in `unify` still persists.

Reply:
I think that test can just be changed to `Basic(S(1), x)`.

Reply:
Alright I made the necessary changes. 

> Basic(S(1), x)

I added this for now. Perhaps later we could add the ability to internally convert `int` instances to `Integer(1)` if passed like in the original test.

Reply:
The problem is that most of the time the args are already Basic so sympifying them would be mostly unnecessary and would slow a lot of things down. It would be better if we had a way of testing this without actually having the checks at runtime.